### DISCUSSION TITLE
Invalidating Cache

### DISCUSSION THREAD

**Discussion by draganbozhinoski** [ZiggyCreatures/FusionCache#199] [General]
Hello,
I would like to ask if there is any way to invalidate caches on context changes(add or update or delete etc). I was going through the docs but did not find anything.

Thank you in advance.

**Comment by jodydonetti** [comment]
Hi @draganbozhinoski , do you mean invalidating single entries or the entire cache?

**Reply by draganbozhinoski** [reply]
Single, let's say automatically when the context state changes. E.g on updating entity

**Reply by jodydonetti** [reply]
You can invalidate each cache entry by calling either `Remove(...)` or `Expire(...)`: the first will effectively remove a cache entry, the second will remove it from the distributed cache and mark it as expired, so it can be maybe used for [fail-safe](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/FailSafe.md) scenarios
Also, in case of a multi-node setup, FusionCache would take care of propagating that to the [distributed cache](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/CacheLevels.md) too (if configured), and also to the other nodes by using the [backplane](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/Backplane.md) (again, if configured).

But it see you are mentioned a "context", maybe you are referring to a DbContext from EF?

**Reply by draganbozhinoski** [reply]
Yes, i was thinking about dbcontext, is there any way to trigger it?

**Reply by jodydonetti** [reply]
No, not that I'm aware of, at least not yet.

**Reply by olliejm** [reply]
You may want to take a look at [EFCoreSecondLevelCacheInterceptor](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/tree/master). You'll need to write an implementation of `IEFCacheServiceProvider` for FusionCache, like an `EFFusionCacheProvider` then:

```cs
services.AddEFSecondLevelCache(options =>
{
    options.UseCustomCacheProvider<EFFusionCacheProvider>();
    // Other options here...
}
```

This setup works for us. The implementation of `EFFusionCacheProvider` is trivial, you can easily derive it from e.g. [the EasyCaching.Core provider](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/blob/master/src/EFCoreSecondLevelCacheInterceptor/EFEasyCachingCoreProvider.cs).

It's worth noting that this probably reverses the way you'd normally expect to work with the cache, and how e.g. the Fusion docs will show utilising the cache. You instead call EFCore/Linq at the higher level, and caching happens underneath via interception, rather than primarily interacting with the cache and passing factories to optionally go to DB.

**Reply by jodydonetti** [reply]
Hi @olliejm , thanks for chipping in!

I know about EFCoreSecondLevelCacheInterceptor and I'm looking at that from some time now, but from what I saw the current roadblock is [this](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/blob/3b53c5e7803290db10f0de9376f370704e0b9416/src/EFCoreSecondLevelCacheInterceptor/IEFCacheServiceProvider.cs#L11), which is not possible with FusionCache since there's no native `Clear()` method available (for good reasons).

I'm thinking about alternatives, but for now I haven't found any, do you happen to have any idea?

Thanks!

**Reply by olliejm** [reply]
Hey, yeah you are correct - I encountered the missing `Clear()` and assumed it was missing for a reason. 

For now I ignored it, after analysing [the EasyCaching.Core provider's `Clear()` implementation](https://github.com/VahidN/EFCoreSecondLevelCacheInterceptor/blob/master/src/EFCoreSecondLevelCacheInterceptor/EFEasyCachingCoreProvider.cs#L107) and seeing that execution is skipped for the EasyCaching hybrid caching mode. 

With Fusion also being a hybrid cache, I assumed the reason for missing `Clear()` is related to the specifics of hybrid caching, and that given the official EasyCaching provider skips in that case, that it would also be OK (or at least not drastically bad) for us to skip. 

So our implementation is a no-op currently. 

I'm aware I made a bunch of potentially bad assumptions there, but so far we didn't encounter any obvious problems. You may well have a better picture than me about whether those assumptions are sound or not.

I'll take a look at related issues regarding `Clear()`.

**Reply by jodydonetti** [reply]
Hi @olliejm , I think we'll have news about an `EFCoreSecondLevelCacheInterceptor` impl after [v2](https://github.com/ZiggyCreatures/FusionCache/releases/tag/v2.0.0-preview-4) will come out soon 😬 (it will include both [Tagging](https://github.com/ZiggyCreatures/FusionCache/issues/319) and [Clear](https://github.com/ZiggyCreatures/FusionCache/issues/331) support).
